libmamba/src/core/transaction_context.cpp

```diff
@@ -3,13 +3,37 @@
 // Distributed under the terms of the BSD 3-Clause License.
 //
 // The full license is in the file LICENSE, distributed with this software.
+#ifndef _WIN32
+#include <csignal>
+#endif
+
+#include <reproc++/drain.hpp>
 
 #include "mamba/core/transaction_context.hpp"
 #include "mamba/core/output.hpp"
 #include "mamba/core/util.hpp"
 
 namespace mamba
 {
+    void compile_python_sources(std::ostream& out)
+    {
+        out << "from compileall import compile_file\n";
+        out << "from concurrent.futures import ProcessPoolExecutor\n";
```

Comment:
Hmm, this was introduced in python 3.2

I am not sure we should break compatibility with older pythons, especially 2.7 might still be used (at least to recreate historical envs).

Reply:
Definitely (I'm still forced to Python 2.7 myself).

I have a fallback to use `-m compileall` for `Python < 3.5` (basically the same condition that used to be used for adding `-j0` before).

Worst case scenario is that the pyc files fail to be installed and will be generated on first import (most of the time at least, permissions or read only filesystems might mess things up).

Reply:
Why not simply use `multiprocessing`?

I wonder if we should add batching because the number of pyc files may easily grow to 100k and this will create as many jobs as pyc files.

Reply:
We might also want to expose this Python code and not hide it here in the C++ code so that people can use it outside the install process.

Reply:
> Why not simply use multiprocessing?

I started from the [implementation in cpython](https://github.com/python/cpython/blob/ee0ac328d38a86f7907598c94cb88a97635b32f8/Lib/compileall.py#L102-L113) which uses `concurrent.futures` and saw not reason not to use it.

> I wonder if we should add batching because the number of pyc files may easily grow to 100k and this will create as many jobs as pyc files.

With [`max_workers=None`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor) `ProcessPoolExecutor` uses the number of cores and takes care of the batching for sending data the subprocesses.

> We might also want to expose this Python code and not hide it here in the C++ code so that people can use it outside the install process. We could even make it compatible to the command-line interface of compileall

It already exists in cpython however it:
* Waits for stdin to be closed before compilation starts
* It only parallelises when recursing into directories

I'm inclined to improve upstream rather than focusing on it to much here so that `-m compileall` can be used for Python 3.11+ if they approve of the change.

Reply:
Yes that sounds like a much better plan

Reply:
FYI with `multiprocessing` you don't get the batching for free, you'll have to do it manually.